Improve: add pausing, rewind and stopping to replays

SlySven opened the pull request:
At the request of @weisluke (**@Adael/Bludlust** in WoTMUD Discord) I've made some improvements to the replay system:

* Restructure the replay system to hold all the replay chunks in memory and to pick them out as required during a replay.

* Removes the separate timer for replay time display by inserting dummy chunks into the above replay data which have an integer number of seconds as a time but no data. The time information is sent to the display in the replay toolbar as it is read and since there is a real or dummy chunk at least every simulated replay second it automatically keeps in step with the replay elapsed time - however fast it is being replayed.

* A pause/resume control has been added which stops (and restarts with an appropriate interval) the timer used to provide the delay between replaying each chunk.

* A rewind control has been added which pauses the replay and then rewinds things to the start ready to be re-run by clicking the resume button.

* An abort (eject) has been added which immediately terminates the replay and resets the application to run another one.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Improve: allow slower than normal replays, and pausing, rewind and stopping

Restructure the replay system to hold all the replay chunks in memory
and to pick them out as required during a replay.

Removes the separate timer for replay time display by inserting dummy
chunks into the above replay data which have an integer number of seconds
as a time but no data. The time information is sent to the display in the
replay toolbar as it is read and since there is a real or dummy chunk at
least every simulated replay second it automatically keeps in step with
the replay elapsed time - however fast it is being replayed.

A pause/resume control has been added which stops the timer used to provide
the delay between replaying each chunk.

A rewind control has been added which pauses the replay and then rewinds
things to the start ready to be re-run by clicking the resume button.

A replay abort (eject) has been added which immediately terminates the
replay and resets the application to run another one.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=6474&src=https%3A%2F%2Fgithub.com%2FMudlet%2FMudlet">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=6474&src=https%3A%2F%2Fgithub.com%2FMudlet%2FMudlet)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/Mudlet/Mudlet/6474/9b4ca8c3/d3f67b582e1a68efc11035ac00ba217ed13a3d8a.svg)](https://app.codesee.io/r/reviews?pr=6474&src=https%3A%2F%2Fgithub.com%2FMudlet%2FMudlet)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

SlySven committed:
Fixup: remove extraneous white space that made CodeFactor unhappy

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: correct error in calculating timing intervals between replay chunks

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

weisluke commented:
Did a bit of playing around with this. @SlySven fixed the one error I picked up on regarding some of the timing, so as of right now it appears to be working fabulously! Wonderful work, really appreciated :)

vadi2 reviewed (commented):
This is really cool!

Two comments...

> Restructure the replay system to hold all the replay chunks in memory and to pick them out as required during a replay.

That's going to make it tough for memory-constrained devices (like pi's) to handle multi-hour replays from things like massive in-game events. It would be a shame to lose this ability, I think.

Also, I tried recording a sample replay and playing it back, but the playback does not quite work - instead of showing the text piece by piece as it came in, it shows it all at once: 
[replay.zip](https://github.com/Mudlet/Mudlet/files/10327130/replay.zip)

SlySven committed:
Fix: make end of replay message appear AFTER the last replay chunk is used

Otherwise the message is confusing...!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> That's going to make it tough for memory-constrained devices (like pi's) to handle multi-hour replays from things like massive in-game events. It would be a shame to lose this ability, I think.

Well, it might be a swings and roundabouts situation - we need to read through the file once (or twice in the faulty case) before starting the replay now to check/determine whether it was a faulty one from the period where the wrong size integer (`qint64` instead of `qint32`) was used for the interval from the previous chunk. This PR takes that opportunity to read the file into memory - which could be significant when the replay speed is well above unity because the original code would be reading the file from disk whereas now the code will read from, presumably, faster memory.

Why don't we give it a shot and revisit the original code if it really does become an issue. It may be that a compile time option could be used (for low memory cases) that would, instead of storing the data in memory, keep a file offset for the first data byte and use that instead. This could be slightly awkward because ideally/theoretically we would want to lock the file to ensure it does not change between checking it and then actually using it. However I am not wildly keen to code that alternative unless it is actually proven to be necessary.

> Also, I tried recording a sample replay and playing it back, but the playback does not quite work - instead of showing the text piece by piece as it came in, it shows it all at once:
[replay.zip](https://github.com/Mudlet/Mudlet/files/10327130/replay.zip)

I've just tried it with the most recent commit cde4583 and it worked fine for me.  Can you recheck it. It is possible that the last chunk was not handled correctly and I see that the replay only lasted for 9.653 seconds and only had twelve chunks - and all the text in white with underlined links (where the cursor is) and the following text appeared towards the end (the time-stamps relate to when the text was parsed internally so do not quite correspond to when they get painted on screen I think) - that cursor position was reached after about 4 seconds on the timer then there was a pause for around another 4 seconds before the remainder appeared:
![image](https://user-images.githubusercontent.com/6163092/210152423-a6d5f6f4-05fc-4fa7-9232-f91b0eff71af.png)

weisluke commented:
I just reproduced Vadi's issue on cde4583 (for Windows at least), in the sense that it doesn't show up on the screen until all at once, even though the timestamps show differences. 

https://user-images.githubusercontent.com/52049495/210153027-be707b23-1e64-4b77-aeaf-94e4c35f9f27.mp4

weisluke commented:
FWIW, recording a replay on WoTMUD and replaying it seems to work fine (although the issue with the last line of the replay showing up after the "The replay has ended." message still seems to be there). 

https://user-images.githubusercontent.com/52049495/210153224-aa2c9933-14f9-4cab-a175-cebfc119a71f.mp4

SlySven commented:
I'm not exactly sure what is going on but it seems that if the `TConsole` is full up (i.e. to the bottom of the screen) it is not scrolling up when the new content would be printed out (although trying to scroll it with the {vertical} mouse wheel does allow it to be brought into view. If. with that sample you slow it down to x0,25 and allow it to reach 0:00:04 and then pause it - there will have been some text sent to the screen (after around the 0:00:02 point) but not shown. but you can reveal it by scrolling.

Ironically the ability to slow things down and to rewind them has actually greatly aided me in finding this - let's hope I can nail what is not working quickly...!

SlySven committed:
Revise: fix failure to scroll up & show new lines off the bottom of screen

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_replayUIAndAllowSlowerThanNormalReplays

SlySven committed:
Merge branch 'development' into Improve_replayUIAndAllowSlowerThanNormalReplays

SlySven commented:
~~Deleted~~ *I made a duplicate post here of a previous one.* :man_facepalming:

SlySven committed:
Merge branch 'development' into Improve_replayUIAndAllowSlowerThanNormalReplays

vadi2 commented:
Sounds fine, could you test how a 3-4 hour replay holds up so we're not going in blind with this change? We have a pi to test this on as well

SlySven commented:
Have you got one I can test - my longest is a couple of hours and it is only around a quarter of a megabyte:
![image](https://user-images.githubusercontent.com/6163092/210875187-8d2395d5-5465-4ec9-a1ea-ff5a4098c08f.png)

SlySven committed:
Merge branch 'development' into Improve_replayUIAndAllowSlowerThanNormalReplays

Conflicts resolved in:
* src/mudlet.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

weisluke commented:
@SlySven I have noticed one odd bug with the replays when playing around. I played back the replay Vadi attached early on. Then, I tried playing a replay of mine from WoTMUD. The lines of the WoTMUD replay again came in sporadically/slower than anticipated.
If I restart Mudlet and then ONLY play the WoTMUD replay, everything is fine. Honestly kind of hard to describe (I know what I've written here probably isn't the best :P), better to see in real time, but the screen recording is too large to upload here. Fine to link you to dropbox or something if I upload them elsewhere?

SlySven commented:
> Fine to link you to dropbox or something if I upload them elsewhere?

You can post the link here if the replay does not contain sensitive information. Or PM me in Discord if it does.

The only things I can think of are:
* the replay from Vadi does contain MXP stuff which might impact on a profile that is not normally used with it (as there must be a small additional processor burden) if the replay introduces elements that have to be checked for/handled in following data. 
* the profile can be modified during the processing of a replay, and it is/can not be reset afterwards - so if a replay changes things (like an automatically adjusting tic-timer that gets set to a shorter/longer interval because the replay is being played at a speed other than "real-time" {**X1**}) then those will be "off" afterwards.

weisluke commented:
Hey @SlySven,

Posted two files here: https://www.dropbox.com/sh/f3woxfngy2wo3l4/AADmbBPJAZFV2G-Gkoyc2QOha?dl=0
The smaller one (I'll call it file 1) is me replaying just a snippet from WoTMUD. 
The larger one (I'll call it file 2) is replaying the file from Vadi, then the same replay from WoTMUD as in file 1.

Main difference is in how the prompt lines from the WoTMUD replay appear in file 2, when compared to file 1.

Let me know if you have any questions about them!

SlySven committed:
Merge branch 'development' into Improve_replayUIAndAllowSlowerThanNormalReplays

SlySven committed:
Fix: correct an off-by-one error in assembling the replay data

The very last second before the end of the replay was not updating the
displayed elapsed time in the Replay Toolbar.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_replayUIAndAllowSlowerThanNormalReplays

vadi2 commented:
@SlySven  I'll close this for now as the PR hasn't had activity for a long while. Please re-open when it is ready.
